stan/math/rev/mat/fun/multiply.hpp

```diff
@@ -1,23 +1,407 @@
+                       const Eigen::Matrix<TB, CA_, 1>& B)
+        : vari(0.0),
+          CA(A.cols()),
+          Ad_a(ChainableStack::memalloc_.alloc_array<double>
+               (A.cols())),
+          Bd_a(ChainableStack::memalloc_.alloc_array<double>
+               (B.rows())),
+          variRefA_(ChainableStack::memalloc_.alloc_array<vari*>
+                    (A.rows() * A.cols())),
+          variRefB_(ChainableStack::memalloc_.alloc_array<vari*>
+                    (B.rows() * B.cols())) {
+            for (size_type i = 0; i < A.size(); ++i) {
+                variRefA_[i] = A.coeffRef(i).vi_;
+                Ad_a[i] = A.coeffRef(i).val();
+            }
+            for (size_type i = 0; i < B.size(); ++i) {
+                variRefB_[i] = B.coeffRef(i).vi_;
+                Bd_a[i] = B.coeffRef(i).val();
+            }
+            double AB = Eigen::Map<
+              Eigen::Matrix<double, 1, CA_> > (Ad_a, 1, CA)
+                        * Eigen::Map<
+              Eigen::Matrix<double, CA_, 1> > (Bd_a, CA, 1);
+            variRefAB_ = new vari(AB, false);
+        }
+
+      virtual void chain() {
+        using Eigen::Matrix;
+        using Eigen::Map;
+        double adjAB;
+        Matrix<double, 1, CA_> adjA(1, CA);
+        Matrix<double, CA_, 1> adjB(CA, 1);
+
+        adjAB = variRefAB_->adj_;
+        adjA = adjAB * Map<Matrix<double, CA_, 1> >(Bd_a, CA, 1).transpose();
+        adjB = Map<Matrix<double, 1, CA_> >(Ad_a, 1, CA).transpose() * adjAB;
+        for (size_type i = 0; i < CA; ++i)
+            variRefA_[i]->adj_ += adjA(i);
+        for (size_type i = 0; i < CA; ++i)
+            variRefB_[i]->adj_ += adjB(i);
+      }
+    };
+
+    template <int RA_, int CA_, typename TB, int CB_>
+    class multiply_mat_vari<double, RA_, CA_, TB, CB_> : public vari {
+    public:
+      int RA, CA, CB, Asize, Bsize;  // A.cols() = B.rows()
+      double* Ad_a;
+      double* Bd_a;
+      vari** variRefB_;
+      vari** variRefAB_;
+
+      /* ctor for multiply function
+       *
+       * Stores pointers to varis for B
+       * Stores pointers to varis for AB
+       * All memory allocated on ChainableStack
+       *
+       * @tparam RA_ Rows for matrix A
+       * @tparam CA_ Columns for matrix A, Rows for matrix B
+       * @tparam TB Scalar type for matrix B
+       * @tparam CB_ Columns for matrix B
+       * @param matrix A
+       * @param matrix B
+       * */
+     multiply_mat_vari(const Eigen::Matrix<double, RA_, CA_>& A,
+                       const Eigen::Matrix<TB, CA_, CB_>& B)
+        : vari(0.0),
+          RA(A.rows()), CA(A.cols()),
+          CB(B.cols()), Asize(A.size()), Bsize(B.size()),
+          Ad_a(ChainableStack::memalloc_.alloc_array<double>
+               (A.rows() * A.cols())),
+          Bd_a(ChainableStack::memalloc_.alloc_array<double>
+               (B.rows() * B.cols())),
+          variRefB_(ChainableStack::memalloc_.alloc_array<vari*>
```

Comment:
Prefer `B.size()` to `B.rows() * B.cols()` --- then it'll fit on one line, too.  I really don't like starting the arguments on their own line.  The open parent should go on the previous line in any case if you do this.   cpplint doesn't pick up everything in the Google style guide.